doc/api/diagnostics_channel.md

```diff
@@ -409,6 +459,534 @@ channel.subscribe(onMessage);
+the task "completes". In the case of a synchronous task the `result` will be
+the return value and the `error` will be anything thrown from the function.
+With callback-based async functions the `result` will be the second argument
+of the callback while the `error` will either be a thrown error visible in the
+`end` event or the first callback argument in either of the `asyncStart` or
+`asyncEnd` events.
+
+Tracing channels should follow a naming pattern of:
+
+* `tracing:module.class.method:start` or `tracing:module.function:start`
+* `tracing:module.class.method:end` or `tracing:module.function:end`
+* `tracing:module.class.method:asyncStart` or `tracing:module.function:asyncStart`
+* `tracing:module.class.method:asyncEnd` or `tracing:module.function:asyncEnd`
+* `tracing:module.class.method:error` or `tracing:module.function:error`
+
+#### `start(event)`
+
+* Name: `tracing:${name}:start`
+
+The `start` event represents the synchronous start of a traceable function.
+At this point the event data may contain function arguments or anything else
+available at the very start of the execution of the function.
+
+#### `end(event)`
+
+* Name: `tracing:${name}:end`
+
+The `end` event represents the synchronous end of a traceable function. At this
+point, if the traced function was synchronous the `result` field will be set to
+the return value of the function. Alternatively, the `error` field may be
+present to represent any thrown errors.
+
+It is recommended to listen specifically to the `error` event to track errors
+as it may be possible for a traceable action to produce multiple errors. For
+example, an async task which fails may be started internally before the sync
+part of the task then throws an error.
+
+#### `asyncStart(event)`
+
+* Name: `tracing:${name}:asyncStart`
+
+The `asyncStart` event represents the callback or continuation of a traceable
+function being reached. At this point things like callback arguments may be
+available, or anything else expressing the "result" of the action.
+
+For callbacks-based functions, the first argument of the callback will be
+assigned to the `error` field, if not `undefined` or `null`, and the second
+argument will be assigned to the `result` field.
+
+For promises, the argument to the `resolve` path will be assigned to `result`
+or the argument to the `reject` path will be assign to `error`.
+
+It is recommended to listen specifically to the `error` event to track errors
+as it may be possible for a traceable action to produce multiple errors. For
+example, an async task which fails may be started internally before the sync
+part of the task then throws an error.
+
+#### `asyncEnd(event)`
+
+* Name: `tracing:${name}:asyncEnd`
+
+The `asyncEnd` event represents the callback of an asynchronous traceable
+function being completed. It's not likely event data will change after the
+`asyncStart` event, however it may be useful for some tracers to see the point
```

Comment:
The term `tracer` probably also needs definition, or replacement with something clearer.